File: data_structures/c/stack/stack.c
#include "stack.h"

#include <stdlib.h>


static node_t* create_node(item_t *item) {
    node_t *node = (node_t *) malloc(sizeof(node_t));
    node->item = item;
    node->next = NULL;

    return node;
}

lstack_t* create_stack(int size) {
    lstack_t *stack = (lstack_t *) malloc(sizeof(lstack_t));
    stack->size = size;
    stack->top = NULL;
    stack->length = 0;

    return stack;
}

int push_on_stack(lstack_t *stack, item_t *item) {
    if(stack->size == 0 || stack->length < stack->size) {
        node_t *new_node = create_node(item);
        new_node->next = stack->top;
        stack->top = new_node;
        stack->length++;
        return 1;
    }

    return 0;
}

item_t* pop_on_stack(lstack_t *stack) {
    item_t *item = NULL;
    if(stack->top) {
        node_t *removed = stack->top;
        stack->top = removed->next;
        item = removed->item;
        stack->length--;
        free(removed);
    }
    return item;
}

item_t* peek_on_stack(lstack_t *stack) {
    if(stack->top) {
        return stack->top->item;
    }
    return NULL;
}

File: data_structures/go/stack/stack.go
package stack

import "errors"

// Node structure
type Node struct {
	item *Item
	next *Node
}

// Stack main structure
type Stack struct {
	top    *Node
	size   int
	length int
}

var EmptyStackError = errors.New("Stack is empty")
var FullStackError = errors.New("Stack is full")

// Create a new empty stack
func CreateStack(size int) *Stack {
	return &Stack{
		top:    nil,
		size:   size,
		length: 0,
	}
}

// Create a new Node
func CreateNode(item *Item) *Node {
	return &Node{
		item: item,
		next: nil,
	}
}

// Push an item into stack, Returns error if stack is full
func (stack *Stack) Push(item *Item) error {
	if stack.size == 0 || stack.length < stack.size {
		new_node := CreateNode(item)
		new_node.next = stack.top
		stack.top = new_node
		stack.length++
		return nil
	}
	return FullStackError
}

// Pop an item from stack. Returns error if stack is empty
func (stack *Stack) Pop() (*Item, error) {
	if stack.top == nil {
		return nil, EmptyStackError
	}
	removed_node := stack.top
	stack.top = removed_node.next
	item := removed_node.item
	stack.length--

	return item, nil
}

// Peek an item from stack withour removing. Returns error if stack is empty
func (stack *Stack) Peek() (*Item, error) {
	if stack.top == nil {
		return nil, EmptyStackError
	}
	return stack.top.item, nil
}
